File: leetcode/python/P0077_Combinations.py
# https://leetcode.com/problems/combinations/

from copy import copy
from typing import List


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        result = []

        def backtrack(
            n: int,
            k: int,
            crt: int,
            val: int,
            crt_vec: List[int],
        ) -> None:
            if crt == k:
                result.append(copy(crt_vec))
                return

            for v in range(val, n + 1):
                crt_vec[crt] = v
                backtrack(n, k, crt + 1, v + 1, crt_vec)

        crt_vec = [0] * k
        backtrack(n, k, 0, 1, crt_vec)
        return result

File: leetcode/cpp/P0077_Combinations.cpp
// https://leetcode.com/problems/combinations/

#include <iostream>
#include <vector>

class Solution {
 public:
  std::vector<std::vector<int>> combine(int n, int k) {
    std::vector<std::vector<int>> result;
    std::vector<int> crt_vec(k, 0);
    backtrack(n, k, 0, 1, crt_vec, result);
    return result;
  }

  void backtrack(int n, int k, int crt, int val, std::vector<int>& crt_vec,
                 std::vector<std::vector<int>>& result) {
    if (crt >= k) {
      result.push_back(crt_vec);
      return;
    }

    for (int ii = val; ii <= n; ++ii) {
      crt_vec[crt] = ii;
      backtrack(n, k, crt + 1, ii + 1, crt_vec, result);
    }
  }
};

int main(int argc, char** argv) {
  Solution solution;
  std::vector<std::vector<int>> result = solution.combine(4, 2);

  for (const auto& vec : result) {
    for (const auto& val : vec) {
      std::cout << val << " ";
    }
    std::cout << std::endl;
  }

  return 0;
}